syntax = "proto3";

service AnsibleExecutor {
   rpc StartSession(EmptyMsg) returns (StartSessionMSG) {}
   rpc InitExecution(Config) returns (Status) {}
   rpc RunTask(Task) returns (TaskResults) {}
   rpc StopExecution(SessionID) returns (Status) {}
   // rpc RestartExecution(SessionID) returns (Status) {}
}

message StartSessionMSG {
   bool ok = 1;
   string ID = 2;
   string error_msg = 3;
}

message SessionID {
   string session_ID = 1;
}

message EmptyMsg {}

message MapFieldEntry {
   string key = 1;
   string value = 2;
}

message Config {
   string session_ID = 1;
   string pb_path = 2;
   string inv_path = 3;
   string emty_task_name = 4;
   repeated MapFieldEntry extra_vars = 5;
   repeated MapFieldEntry env_vars = 6;
}

message Task {
   string session_ID = 1;
   string task_str = 2;
}

message TaskResult {
   bool ok = 1;
   string results_dict_str = 2;
   string task_name = 3;
   bool is_changed = 4;
   bool is_failed = 5;
   bool is_skipped = 6;
   bool is_unreachable = 7;
   bool is_ignored_errors = 8;
   bool is_ignored_unreachable = 9;
   string stdout = 10;
   string stderr = 11;
   string msg = 12;
}

message TaskResults {
   bool task_adding_ok = 1;
   string task_adding_error = 2;
   repeated TaskResult task_results = 3;
}

message Status {
   bool ok = 1;
   string error_msg = 2;
}